### DISCUSSION TITLE
AddLinPoly example (Python)

### DISCUSSION THREAD

**Discussion by VolkerMuehlhaus** [thliebig/openEMS-Project#56] [Q&A Usage] [answered]
Hi all,

I am struggling with the polygon point syntax of **AddLinPoly for Python interface**. For OpenEMS in Matlab syntax, there is an example and that works well. But I don't understand how to create the polygon point variable for AddLinPoly in Python.

Could someone help me out with an example? Something simple with 4 points would be fine, just to see how the variable is created and populated with data and passed to the AddLinPoly call.

Best regards
Volker

PS: This is what I tried:

~~~~~~~~~~~ snip ~~~~~~~~~~~~~~

pts_x = np.array([])
pts_y = np.array([])

pts_x = r_[pts_x, -100]
pts_y = r_[pts_y, -100]

pts_x = r_[pts_x, 100]
pts_y = r_[pts_y, -100]

pts_x = r_[pts_x, 100]
pts_y = r_[pts_y, 100]

pts_x = r_[pts_x, 100]
pts_y = r_[pts_y, -100]

pts = np.array([pts_x, pts_y])

TopMetal2.AddLinPoly(pts, 'z', 300, 10)

~~~~~~~~~~~ snip ~~~~~~~~~~~~~~

but in CSXCAD this gives error messages and the polygon doesn't show up:

ERROR: In C:\Users\liebig\Downloads\VTK-8.2.0\Filters\Modeling\vtkLinearExtrusionFilter.cxx, line 124
vtkLinearExtrusionFilter (000001DC36ADBC50): No data to extrude!

ERROR: In C:\Users\liebig\Downloads\VTK-8.2.0\Filters\General\vtkTransformPolyDataFilter.cxx, line 85
vtkTransformPolyDataFilter (000001DC36B34240): No input data

**Comment by thliebig** [comment]
Well this is why openEMS is open source, you can always have a look in how it's implemented. As I did not remember myself, I had to do it too... See [here](https://github.com/thliebig/CSXCAD/blob/master/python/CSXCAD/CSPrimitives.pyx#L722)

But looking at it it seems you did everything correctly? Maybe just nobody has tried a polygon in python yet?

**Reply by thliebig** [reply]
You could use  "GetCoords" from the resulting polygon to see if setting the points did indeed work as it should...

**Reply by VolkerMuehlhaus** [reply]
This is what I get with print() of GetCoords:
(array([-100.,  100.,  100.,  100.]), array([-100., -100.,  100., -100.]))

Based on my (limited) understanding of the expected data, that looks fine.
I also queried the other parameters and they all match.

The corresponding XML output looks like this:
`<LinPoly Priority="0" Elevation="3.000000e+02" NormDir="2" QtyVertices="4" Length="1.000000e+01">
<Vertex X1="-1.000000e+02" X2="-1.000000e+02"/>
<Vertex X1="1.000000e+02" X2="-1.000000e+02"/>
<Vertex X1="1.000000e+02" X2="1.000000e+02"/>
<Vertex X1="1.000000e+02" X2="-1.000000e+02"/>
</LinPoly>`

**Reply by VolkerMuehlhaus** [reply, marked as answer]
Problem solved, it was a foolish mistake with my coordinates: there was a typo with the sign of one entry, resulting in invalid shape. The call itself was correct. Thanks for guiding me to find my mistake!

**Reply by thliebig** [reply]
Happy that you found it. Maybe for debugging you could just use matplotlib?
Something like plt.plot(pts[0], pts[1]) ?

**Reply by VolkerMuehlhaus** [reply]
Good idea, thanks Thorsten!

My invalid "handmade" polygons were just for testing the syntax. The "real" polygon data comes from GDSII files. For OpenEMS in Matlab syntax, I already created a Python script to convert GDSII layout to OpenEMS polygons, and now derive another version to create OpenEMS Python syntax. This is why I checked the syntax to be created by my GDSII translator.
